deltachat-ffi/src/lib.rs

```diff
@@ -86,12 +86,62 @@ pub unsafe extern "C" fn dc_context_new(
     match ctx {
         Ok(ctx) => Box::into_raw(Box::new(ctx)),
         Err(err) => {
-            eprintln!("failed to create context: {}", err);
+            eprintln!("failed to create context: {:#}", err);
             ptr::null_mut()
         }
     }
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn dc_context_new_closed(dbfile: *const libc::c_char) -> *mut dc_context_t {
+    setup_panic!();
+
+    if dbfile.is_null() {
+        eprintln!("ignoring careless call to dc_context_new_closed()");
+        return ptr::null_mut();
+    }
+
+    let id = rand::thread_rng().gen();
+    match block_on(Context::new_closed(
+        as_path(dbfile).to_path_buf().into(),
+        id,
+    )) {
+        Ok(context) => Box::into_raw(Box::new(context)),
+        Err(err) => {
+            eprintln!("failed to create context: {:#}", err);
+            ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn dc_context_open(
+    context: *mut dc_context_t,
+    passphrase: *const libc::c_char,
+) -> libc::c_int {
+    if context.is_null() {
+        eprintln!("ignoring careless call to dc_context_open()");
+        return 0;
+    }
+
+    let ctx = &*context;
+    let passphrase = to_string_lossy(passphrase);
+    block_on(ctx.open(passphrase))
+        .log_err(ctx, "dc_context_open() failed")
+        .is_ok() as libc::c_int
```

Comment:
The doc string says: `return 1 if the database is opened with this passphrase, 0 if the passphrase is incorrect or on error.`

But with `is_ok()` we are only returning whether there was an error, and `1` if the passphrase was incorrect.

--

Also it would be nice if the UI can show the user whether the passphrase was wrong or there was another error, so, e.g.

```suggestion
        .map(|b| b as libc::c_int)
        .unwrap_or(2)
```

 and `return 1 if the database is opened with this passphrase, 0 if the passphrase is incorrect, 2 on error.`

Reply:
I fixed the bug, but made it return `0` on errors.

I am not sure about returning `2`. What the UI should do if `2` is returned? `0` means we failed to open the database, `1` means we opened the database. The error is logged so you can display it in a toast, but the end result is the same: the database is not open.

We already return default value on errors in many places, for example in `dc_get_draft` and `dc_get_msg_cnt`, and don't use special values like `2` or `-1` for errors in the API. This way we avoid having untested code paths in the UI code that should display errors that almost never happen.

If opening databases happens in the background without explicit user action, there may be no place to display the error other than the log.

Reply:
> What the UI should do if `2` is returned?

Show "Error decrypting database" instead of "Wrong passphrase" - obviously only if the user explicitly entered a passphrase.

But it's ok to not know it and always show some generic error message (or simply ask for the passphrase again, like the Firefox password manager), and of course you're right about untested code paths.